### PR TITLE
Fix new warning STL4037 raised by VS2022

### PR DESCRIPTION
**Pull request by dzenanz**
Example error message:

`M:\Dashboard\ITK\Modules\Core\Common\src\itkNumericTraits.cxx(77,85): warning C4996: 'std::complex<char>::complex': warning STL4037: The effect of instantiating the template std::complex for any type other than float, double, or long double is unspecified. You can define _SILENCE_NONFLOATING_COMPLEX_DEPRECATION_WARNING to suppress this warning. [M:\Dashboard\ITK-build\Modules\Core\Common\src\ITKCommon.vcxproj]`

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)

### CODE REVIEW

**Comment by dzenanz** [comment]
After doing this, I realized it might have been better to have put integral complex types under legacy ifdefs. What do think? The same thing exists in VXL, at `Modules/ThirdParty/VNL/src/vxl/core/vnl/vnl_complex_traits.h`.

**Comment by dzenanz** [comment]
I changed one of my builds to use VS2022 today: [Windows11-VS22x64-RelWithDebInfo-Favorite-Remotes](https://open.cdash.org/build/8296517).

File: Modules/Filtering/ImageFrequency/test/itkFrequencyFFTLayoutImageRegionIteratorWithIndexTest.cxx [hunk 1]
```diff
@@ -263,7 +263,7 @@ itkFrequencyFFTLayoutImageRegionIteratorWithIndexTest(int, char *[])
   constexpr unsigned int Dimension = 3;
 
   using CharPixelType = char;
-  using FloatPixelType = char;
+  using FloatPixelType = float;
```
Comment (jhlegarreta) [inline Modules/Filtering/ImageFrequency/test/itkFrequencyFFTLayoutImageRegionIteratorWithIndexTest.cxx:266]: I'd dare to say that this change is unrelated to the `std::complex` warning. If that is the case, can it please be put into a separate commit so that the corresponding warning is clear, or even better, a separate PR so that it can be merged independently?

**Review by jhlegarreta — COMMENTED** [review]
Looks good to me. I'd prefer the FFT change to be put into a separate commit/PR if it is unrelated to the `std::complex` warning. Thanks.

File: Modules/Core/Common/include/itkNumericTraits.h [hunk 2]
```diff
@@ -1979,6 +1980,13 @@ class NumericTraits<std::complex<TComponent>>
     }
     m = NumericTraits<ValueType>::ZeroValue();
   }
+
+#ifdef ITK_USE_CONCEPT_CHECKING
+  static_assert(std::is_same<TComponent, float>::value || std::is_same<TComponent, double>::value ||
+                  std::is_same<TComponent, long double>::value,
+                "As per https://en.cppreference.com/w/cpp/numeric/complex the behavior is unspecified and may fail to "
+                "compile if TComponent is not float, double, or long double and undefined if T is not NumericType.");
```
Comment (N-Dekker) [inline Modules/Core/Common/include/itkNumericTraits.h:1988]: OK, but I'd rather just do `static_assert(std::is_floating_point<TComponent>::value, "...")`

File: Modules/Core/Common/test/itkNumericTraitsTest.cxx [hunk 3]
```diff
@@ -364,12 +350,17 @@ CheckIsComplexTraits()
                 "std::complex<float> does not have the correct IsComplex trait");
   static_assert(itk::NumericTraits<std::complex<double>>::IsComplex,
                 "std::complex<double> does not have the correct IsComplex trait");
+  static_assert(itk::NumericTraits<std::complex<long double>>::IsComplex,
+                "std::complex<long double> does not have the correct IsComplex trait");
```
Comment (N-Dekker) [inline Modules/Core/Common/test/itkNumericTraitsTest.cxx:354]: Should we really care about `long double` support? On MSVC, `long double` is basically just like `double`: 64-bits floating point.

File: Modules/Core/Common/include/itkNumericTraits.h (hunk 2 above)
Comment (dzenanz) [reply Modules/Core/Common/include/itkNumericTraits.h:1988]: Good suggestion, will do.

File: Modules/Core/Common/test/itkNumericTraitsTest.cxx (hunk 3 above)
Comment (dzenanz) [reply Modules/Core/Common/test/itkNumericTraitsTest.cxx:354]: It is rarely used nowadays indeed. But it is still an existing type, so we can keep supporting it.

File: Modules/Filtering/ImageFrequency/test/itkFrequencyFFTLayoutImageRegionIteratorWithIndexTest.cxx (hunk 1 above)
Comment (dzenanz) [reply Modules/Filtering/ImageFrequency/test/itkFrequencyFFTLayoutImageRegionIteratorWithIndexTest.cxx:266]: I put it into a separate commit.

File: Modules/Core/Common/test/itkNumericTraitsTest.cxx (hunk 3 above)
Comment (N-Dekker) [reply Modules/Core/Common/test/itkNumericTraitsTest.cxx:370]: OK, thanks.

**Comment by N-Dekker** [comment]
For my understanding, it the idea to keep supporting `complex<int>`, _unless_ ITK_USE_CONCEPT_CHECKING is enabled?

**Comment by dzenanz** [comment]
Correct. My [question](https://github.com/InsightSoftwareConsortium/ITK/pull/3768#issuecomment-1327869271) was should we switch from `CONCEPT_CHECKING` to `LEGACY`? Or maybe just outright remove it, as its behavior is not well defined per C++ standard.

**Comment by N-Dekker** [comment]
@dzenanz Sorry, I accidentally edited your comment, instead of replying to it!

**Comment by blowekamp** [comment]
I recall wanting to remove them a while ago because the non-real types were undefined. The only really issue would be the "One" and "Zero" symbols in the library. However, the `OneValue` and `ZeroValue` methods are the preferred approach now, so it is not as significant.

I agree that the complex number of integer usage should be LEGACY, and not dependent on concept checking.

**Comment by dzenanz** [comment]
I updated the PR to use the standard `ITK_LEGACY_REMOVE` CMake option.

**Comment by jhlegarreta** [comment]
If I understand well, concept checking will refuse to have integer values represented as complex numbers if left and merged, which may break some people's code. Even if it has unpredictable results, marking it as legacy code seems more appropriate to me, and concept checking adds another layer of protection that we eventually would be willing to remove anyways.

File: Modules/Core/Common/test/itkNumericTraitsTest.cxx [hunk 4]
```diff
@@ -1273,6 +1281,7 @@ itkNumericTraitsTest(int, char *[])
   CheckFixedArrayTraits(std::complex<unsigned int>());
   CheckFixedArrayTraits(std::complex<long>());
   CheckFixedArrayTraits(std::complex<unsigned long>());
+#  endif // !ITK_USE_CONCEPT_CHECKING
```
Comment (jhlegarreta) [inline Modules/Core/Common/test/itkNumericTraitsTest.cxx:1284]: endif // !ITK_LEGACY_REMOVE


Or remove the comment altogether.

**Comment by dzenanz** [comment]
I just created #3775 to keep track of my though about it.

**Comment by dzenanz** [comment]
The latest force-push adds an entry in migration guide about this change.
